Wikipedia/Code/TalkPageTopicComposeViewController+TalkPageFormattingToolbar.swift

```diff
@@ -0,0 +1,42 @@
+import Foundation
+
+extension TalkPageTopicComposeViewController: TalkPageFormattingToolbarViewDelegate {
+
+    func didSelectBold() {
+        bodyTextView.addOrRemoveStringFormattingCharacters(formattingString: "'''", cursorOffset: 3)
+    }
+
+    func didSelectItalics() {
+        bodyTextView.addOrRemoveStringFormattingCharacters(formattingString: "''", cursorOffset: 2)
+    }
+
+    func didSelectInsertLink() {
+        guard let link = Link(page: "", label: "", exists: false) else {
+            return
+        }
+        let insertLinkViewController = InsertLinkViewController(link: link, siteURL: viewModel.siteUrl, dataStore: MWKDataStore.shared())
+        insertLinkViewController.delegate = self
+        let navigationController = WMFThemeableNavigationController(rootViewController: insertLinkViewController, theme: self.theme)
+        self.present(navigationController, animated: true)
+    }
+
+}
+
+extension TalkPageTopicComposeViewController: InsertLinkViewControllerDelegate {
+    func insertLinkViewController(_ insertLinkViewController: InsertLinkViewController, didTapCloseButton button: UIBarButtonItem) {
+        dismiss(animated: true)
+    }
+
+    func insertLinkViewController(_ insertLinkViewController: InsertLinkViewController, didInsertLinkFor page: String, withLabel label: String?) {
+        insertOrEditLink(page: page, label: label)
+        dismiss(animated: true)
+    }
+
+    func insertOrEditLink(page: String, label: String?) {
+        if let label {
```

Comment:
Nice, I didn't know we could do this shorthand!

Reply:
> 1. The formatting buttons on the new topic title textfield are acting a little strange - you can tap a button but the wikitext is added to the body text view. I wonder if we should remove the formatting toolbar for the title entirely for now, especially since we seem to style titles as bold on the native-side by default. Might be a good question for design.

Great catch, I ignored the title field when testing, I removed it from here since it doesn't make sense to allow formatting. 

> 2. I notice on Android's editor (actually on our section editor as well), after highlighting text and applying a bold or italic, it automatically highlights the text again. This actually makes for much quicker toggling of the formatting. This also may be more of a design question, so we can wait to address it in design review phase if you want.
> 
Another good catch, I added the same behavior here.

> 3. For the link wizard, I know we can't do link editing, but for highlighting non-link text, can we have it act a bit more like our section editor? There, the highlighted text is sent into the wizard as a default search term. Then the final wikitext completely replaces the text they originally highlighted. I think that would be a nice bonus that hopefully won't cause you to run into web view / link parsing issues, but let me know if you do.

I'm sending the search term in now (and replacing it with the link syntax once inserted), so it pre-populates the search field, but it still doesn't trigger a search - it requires the user to type in something else. Do you think it is worth the risk of trying to tweak the search controller?
For the editing, I now used logic similar to the one I used on the bold/italics button. There are some limitations to it, for example, if the user selects the term and the label when editing it won't take them to the editing view, but it is working. Maybe that's a limitation we'll have to deal with because of the limitations with wikitext, but I think it's worth testing.

Reply:
>  but it still doesn't trigger a search - it requires the user to type in something else. Do you think it is worth the risk of trying to tweak the search controller?

Ah I see - it looks like that bug exists in our section editor too. I would say that fix is beyond the scope of this task. We can spin off a separate bug task for that to fix in both places.